tests/cpp/test_statement_guard.cpp

```diff
@@ -0,0 +1,54 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <gtest/gtest.h>
+
+#include <fusion.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <statement_guard.h>
+#include <tests/cpp/utils.h>
+#include <tests/cpp/validator.h>
+
+namespace nvfuser {
+
+using StatementGuardTest = NVFuserTest;
+
+TEST_F(StatementGuardTest, ExecuteAfterGuard) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* in = makeContigTensor(2);
+  fusion->addInput(in);
+  TensorView* out = flatten(in);
+  fusion->addOutput(out);
+
+  int64_t num_exprs_before_guard = fusion->numExprs();
+  int64_t num_exprs_in_guard = -1;
+  {
+    FusionGuard fg(fusion.get());
+    StatementGuard sg(fusion.get());
+    add(in, in);
+    num_exprs_in_guard = fusion->numExprs();
+  }
+  int64_t num_exprs_after_guard = fusion->numExprs();
+
+  EXPECT_LT(num_exprs_before_guard, num_exprs_in_guard);
+  EXPECT_GT(num_exprs_in_guard, num_exprs_after_guard);
```

Comment:
Note: because of https://github.com/NVIDIA/Fuser/pull/4111/files#diff-dc5b6ef774cae40b275412f5fa599ee7e3934fcee382ff6cc3e0becb88867d61R19, we can't expect num_exprs_before_guard == num_exprs_after_guard.

Reply:
Could we manually do a `fusion->axioms()` before the guard, and `EXPECT_EQ` here?

Reply:
Yes we can. I chose not to because I didn't want the test to be too implementation dependent.

Reply:
It doesn't seem to make much difference. StatementGuard::StatementGuard already calls axioms(): https://github.com/NVIDIA/Fuser/pull/4111/files#diff-dc5b6ef774cae40b275412f5fa599ee7e3934fcee382ff6cc3e0becb88867d61R19. So the difference would just be explicitly calling axioms() vs initializeAll(). 

Having to call initializeAll in the test is still exposing implementation details. Have you considered to let Fusion::Fusion call axioms()? That will fix my concerns. I'm just not sure how expensive that will be in practice.

Reply:
> Have you considered to let Fusion::Fusion call axioms()? That will fix my concerns. I'm just not sure how expensive that will be in practice.

I do think this makes more sense. And I do have the same concern. Anyway, let's just leave it as is. This is a very minor nitpick.